### DISCUSSION TITLE
Omit the version by providing a configurable default value

### DISCUSSION THREAD

**Discussion by curoky** [go-task/task#744] [Ideas]
Every time when I write a `Taskfile.yaml`, I always need to add `version: '3'` to first line.

I understand the importance of clearly specifying the version number, it can significantly reduce maintenance costs due to functional changes. But is it possible to simplify this step when I know what I'm doing.

A possible solution is to read the default version externally, e.g., 

- from command line args: --use-version=3
- from ENV: TASK_USE_VERSION=3
- from global config file: `~/.config/task/config.yaml`

**Comment by ghostsquad** [comment]
The version is an important part of the schema, and is used not only by the CLI but also by programs, like vscode, that can give you autocompletion and helpful errors/warnings by comparing your local document to this:

https://json.schemastore.org/taskfile.json

Since there are multiple major versions to Task, it would be too ambiguous without this field.

To remove this also violates some principles: it's better to be explicit than implicit. Implicit functionality is very hard to troubleshoot.

Maybe this solution is not the right solution to solve the problem you are encountering. Can you describe in more detail the workflow you have?

**Reply by ssbarnea** [reply]
Yep, keep adding that version entry, it is there for very good reasons and making it optional would break all these use cases. It is a small price to pay for making sure that the config file can evolve.

**Comment by curoky** [comment]
I don't want to cause a quarrel, in my opinion it's just choice, no right or wrong. Just like the json schema mentioned above, once a choice is made, it is difficult to go back. From a user perspective, I created this discussion, but from a developer perspective, obviously this doesn't happen.

In my usage scenario, most scripts are just one or two lines of commands, for example:

- go-task
```
version: '3'

tasks:
  default:
    cmds:
      - echo hello world !
```
Writing the version every time is very annoying, I have used a lot of tools, like make/scons/ninja/cmake/bazel, none of these require a version. Of course I don't think tasks/default/cmds is redundant, instead these will make it clearer.

Anyway go task is a very good tool, the reason why I want to omit the version is because I use it a lot. 

Thanks.

**Reply by ghostsquad** [reply]
I see where you are coming from. Maybe there's a different solution to this problem. Can you walk me through your process to create a new Taskfile? 

What if `task --init` was more helpful? Right now it looks like it this is the file that it creates:

```yaml
# https://taskfile.dev
version: '3'

vars:
  GREETING: Hello, World!

tasks:
  default:
    cmds:
      - echo "{{.GREETING}}"
    silent: true
```

I could imagine that this could be simplified to the bare minimum with:

```yaml
# https://taskfile.dev
version: '3'

tasks:
```

possibly with a command `task --init --bare`. You could alias this to `ti` with `alias ti=task --init --bare`. You could further improve your workflow by automatically opening up this file in your editor of choice:

```
alias ti=task --init --bare && ${VISUAL:-${EDITOR:-vim}} ./Taskfile.yaml
```

I use VSCode a lot, so I set both `VISUAL` and `EDITOR` to `code --wait` in my `.zshrc` file. These are standard environment variables as well, used by other programs. (https://unix.stackexchange.com/questions/4859/visual-vs-editor-what-s-the-difference)

Another option that could be added is `--open`, so you would just have `alias ti=task --init --bare --open` which encapsulates the same logic above.

**Reply by curoky** [reply]
Yeah, I tried it recently, `task init` is a good shortcut, but I think I almost never create files from the command line when i use VSCode 😂.

**Reply by ghostsquad** [reply]
> Yeah, I tried it recently, `task init` is a good shortcut, but I think I almost never create files from the command line when i use VSCode 😂.

https://code.visualstudio.com/docs/editor/userdefinedsnippets

**Reply by curoky** [reply]
> > Yeah, I tried it recently, `task init` is a good shortcut, but I think I almost never create files from the command line when i use VSCode 😂.
> 
> [code.visualstudio.com/docs/editor/userdefinedsnippets](https://code.visualstudio.com/docs/editor/userdefinedsnippets)

This is indeed an elegant solution when using VSCode 🥳. I almost forgot vscode's snippets since using copilot.

```json
{
  "Bootstrap Taskfile": {
    "prefix": "taskfile",
    "body": [
      "version: '3'",
      "",
      "tasks:",
      "  default:",
      "    cmds:",
      "     - $0"
    ],
    "description": "Bootstrap taskfile quickly."
  }
}
```
